gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/handler/webauthn/WebAuthnRegisterHandler.java

```diff
@@ -0,0 +1,252 @@
+            logger.error("Unexpected exception", e);
+            ctx.fail(e);
+        }
+    }
+
+    private void registerV0(RoutingContext ctx) {
+        final JsonObject webauthnRegister = ctx.getBodyAsJson();
+        final Session session = ctx.session();
+
+        // session validation
+        if (session == null) {
+            logger.warn("No session or session handler is missing.");
+            ctx.fail(500);
+            return;
+        }
+
+        if (ctx.user() == null) {
+            logger.warn("User must be authenticated to register WebAuthn credentials.");
+            ctx.fail(401);
+            return;
+        }
+
+        // input validation
+        if (isEmptyString(webauthnRegister, "name") ||
+                isEmptyString(webauthnRegister, "displayName")) {
+            logger.debug("Request missing name or displayName field");
+            ctx.fail(400);
+            return;
+        }
+
+        // get authenticated user
+        User user = ((io.gravitee.am.gateway.handler.common.vertx.web.auth.user.User) ctx.user().getDelegate()).getUser();
+
+        // register credentials
+        webAuthn.createCredentialsOptions(webauthnRegister, createCredentialsOptions -> {
+            if (createCredentialsOptions.failed()) {
+                ctx.fail(createCredentialsOptions.cause());
+                return;
+            }
+
+            final JsonObject credentialsOptions = createCredentialsOptions.result();
+            // force user id with our own user id
+            credentialsOptions.getJsonObject("user").put("id", user.getId());
+
+            // force registration if option is enabled
+            if (domain.getWebAuthnSettings() != null && domain.getWebAuthnSettings().isForceRegistration()) {
+                credentialsOptions.remove("excludeCredentials");
+            }
+
+            // save challenge to the session
+            ctx.session()
+                    .put(ConstantKeys.PASSWORDLESS_CHALLENGE_KEY, credentialsOptions.getString("challenge"))
+                    .put(ConstantKeys.PASSWORDLESS_CHALLENGE_USERNAME_KEY, webauthnRegister.getString("name"));
+
+            ctx.put(ConstantKeys.PASSWORDLESS_ASSERTION, credentialsOptions);
+            ctx.next();
+        });
+    }
+
+    private void registerV1(RoutingContext ctx) {
+        final String assertion = ctx.request().getParam("assertion");
+        if (StringUtils.isEmpty(assertion)) {
+            logger.debug("Request missing assertion field");
+            ctx.fail(400);
+            return;
+        }
+
+        final JsonObject webauthnResp = (JsonObject) Json.decodeValue(assertion);
+        // input validation
+        if (isEmptyString(webauthnResp, "id") ||
+                isEmptyString(webauthnResp, "rawId") ||
+                isEmptyObject(webauthnResp, "response") ||
+                isEmptyString(webauthnResp, "type") ||
+                !"public-key".equals(webauthnResp.getString("type"))) {
```

Comment:
This condition may be moved into WebAuthnHandler as it is also used by the ResponesHandler.
That way input validation is done in the same way whatever id the handler.

Reply:
all this method is duplicated, but I would to keep it like this because WebAuthnResponseHandler will be removed in a next version of AM (probably the 3.19).